quantization/src/encoded_vectors_pq.rs

```diff
@@ -0,0 +1,575 @@
+
+pub struct EncodedVectorsPQ<TStorage: EncodedStorage> {
+    encoded_vectors: TStorage,
+    metadata: Metadata,
+}
+
+/// PQ lookup table
+/// Lookup table is a distance from each query chunck to
+/// each centroid related to this chunk
+pub struct EncodedQueryPQ {
+    lut: Vec<f32>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Metadata {
+    centroids: Vec<Vec<f32>>,
+    vector_division: Vec<Range<usize>>,
+    vector_parameters: VectorParameters,
+}
+
+impl<TStorage: EncodedStorage> EncodedVectorsPQ<TStorage> {
+    /// Encode vector data using product quantization.
+    ///
+    /// # Arguments
+    /// * `data` - iterator over original vector data
+    /// * `storage_builder` - encoding result storage builder
+    /// * `vector_parameters` - parameters of original vector data (dimension, distance, ect)
+    /// * `chunk_size` - Max size of f32 chunk that replaced by centroid index (in original vector dimension)
+    /// * `max_threads` - Max allowed threads for kmeans and encodind process
+    pub fn encode<'a>(
+        data: impl Iterator<Item = &'a [f32]> + Clone + Send,
+        mut storage_builder: impl EncodedStorageBuilder<TStorage> + Send,
+        vector_parameters: &VectorParameters,
+        chunk_size: usize,
+        max_kmeans_threads: usize,
+    ) -> Result<Self, EncodingError> {
+        // first, divide vector into chunks
+        let vector_division = Self::get_vector_division(vector_parameters.dim, chunk_size);
+
+        // then, find flattened centroid positions
+        let centroids_count = 256;
+        let centroids = Self::find_centroids(
+            data.clone(),
+            &vector_division,
+            vector_parameters,
+            centroids_count,
+            max_kmeans_threads,
+        )?;
+
+        // finally, encode data
+        #[allow(clippy::redundant_clone)]
+        Self::encode_storage(
+            data.clone(),
+            &mut storage_builder,
+            &vector_division,
+            &centroids,
+            max_kmeans_threads,
+        )?;
+
+        let storage = storage_builder.build();
+
+        #[cfg(feature = "dump_image")]
+        Self::dump_to_image(data, &storage, &centroids, &vector_division);
+
+        Ok(Self {
+            encoded_vectors: storage,
+            metadata: Metadata {
+                centroids,
+                vector_division,
+                vector_parameters: vector_parameters.clone(),
+            },
+        })
+    }
+
+    pub fn get_quantized_vector_size(
+        vector_parameters: &VectorParameters,
+        chunk_size: usize,
+    ) -> usize {
+        let vector_division = Self::get_vector_division(vector_parameters.dim, chunk_size);
+        vector_division.len()
+    }
+
+    fn get_vector_division(dim: usize, chunk_size: usize) -> Vec<Range<usize>> {
+        (0..dim)
+            .step_by(chunk_size)
+            .map(|i| i..std::cmp::min(i + chunk_size, dim))
+            .collect::<Vec<_>>()
```

Comment:
```suggestion
            .collect()
```

The compiler is smart enough :smile:

Reply:
Compiler is smart enough =)
Done